addons/payment_nuvei/controllers/main.py

```diff
@@ -0,0 +1,121 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+import logging
+import pprint
+
+from werkzeug.exceptions import Forbidden
+
+from odoo import http
+from odoo.exceptions import ValidationError
+from odoo.http import request
+from odoo.tools import consteq
+
+from odoo.addons.payment import utils as payment_utils
+
+_logger = logging.getLogger(__name__)
+
+
+class NuveiController(http.Controller):
+    _return_url = '/payment/nuvei/return'
+    _webhook_url = '/payment/nuvei/webhook'
+    _cancel_url = '/payment/nuvei/cancel'
+
+    @http.route(_return_url, type='http', auth='public', methods=['GET'])
+    def nuvei_return_from_checkout(self, **data):
+        """ Process the notification data sent by Nuvei after redirection.
+
+        :param dict data: The notification data.
+        """
+        _logger.info("Handling redirection from Nuvei with data:\n%s", pprint.pformat(data))
+
+        if data:
+            # Check the integrity of the notification.
+            tx_sudo = request.env['payment.transaction'].sudo()._get_tx_from_notification_data(
+                'nuvei', data
+            )
+            self._verify_notification_signature(data, tx_sudo)
+
+            # Handle the notification data.
+            tx_sudo._handle_notification_data('nuvei', data)
+        else:
+            pass
+        return request.redirect('/payment/status')
+
+    @http.route(_webhook_url, type='http', auth='public', methods=['POST'], csrf=False)
+    def nuvei_webhook(self, **data):
+        """ Process the notification data sent by Nuvei to the webhook.
+
+        See https://docs.nuvei.com/documentation/accept-payment/payment-page/output-parameters.
+
+        :param dict data: The notification data.
+        :return: The 'OK' string to acknowledge the notification.
+        :rtype: str
+        """
+        _logger.info("Notification received from Nuvei with data:\n%s", pprint.pformat(data))
+        try:
+            # Check the integrity of the notification.
+            tx_sudo = request.env['payment.transaction'].sudo()._get_tx_from_notification_data(
+                'nuvei', data
+            )
+            self._verify_notification_signature(data, tx_sudo)
+
+            # Handle the notification data.
+            tx_sudo._handle_notification_data('nuvei', data)
+        except ValidationError:  # Acknowledge the notification to avoid getting spammed.
+            _logger.exception("Unable to handle the notification data; skipping to acknowledge.")
+
+        return 'OK'  # Acknowledge the notification.
+
+    @http.route(
+        _cancel_url, type='http', auth='public', methods=['GET'], csrf=False, save_session=False
+    )
+    def nuvei_return_from_canceled_checkout(self, tx_ref, return_access_tkn, **data):
```

Comment:
And can't we, in fact, use the same base return url for both the cancel and the error, and if we find the tx_ref and return_access_tkn are also there, then we use these instead to search for the tx? It'd avoid some duplicate code and make it easier to understand.